client/electron/vpn_service.ts

```diff
@@ -88,3 +82,68 @@ let statusCb: VpnStatusCallback | undefined = undefined;
 export function onVpnStatusChanged(cb: VpnStatusCallback): void {
   statusCb = cb;
 }
+
+/**
+ * Forwards the VPN connection status change event to the VpnStatusCallback.
+ * @param connJson The JSON string representing the VPNConn interface.
+ */
+function handleVpnConnectionStatusChanged(connJson: string) {
+  const conn = JSON.parse(connJson) as VPNConn;
+  console.debug(`received ${StatusChangedEvent}`, conn);
+  switch (conn?.status) {
+    case VPNConnConnected:
+      statusCb?.(conn.id, TunnelStatus.CONNECTED);
+      break;
+    case VPNConnConnecting:
+      statusCb?.(conn.id, TunnelStatus.RECONNECTING);
+      break;
+    case VPNConnDisconnecting:
+      statusCb?.(conn.id, TunnelStatus.DISCONNECTING);
+      break;
+    case VPNConnDisconnected:
+      statusCb?.(conn.id, TunnelStatus.DISCONNECTED);
+      break;
+  }
+}
+
+// Callback token of the VPNConnStatusChanged event
+let vpnConnStatusChangedCb: CallbackToken | undefined;
+
+/**
+ * Subscribes to all VPN connection related events.
+ * This function ensures that the subscription only happens once.
+ */
+async function subscribeVPNConnEvents(): Promise<void> {
```

Comment:
Pass the VpnStatusCallback here and delete onVpnStatusChanged.

This eliminates the global state and makes the lifecycle/object relationship clearer. Easier to reason about.

Let the app worry about calling it once.

Reply:
Added `InitVpnService` function to `index.ts`, so it have a chance to subscribe to the event only once when the electron app is ready.